### PR TITLE
Use ':' for struct field reference.

### PR DESCRIPTION
**Pull request by hjafarpour**
### Description 
In order to make sure there is no ambiguity in referencing struct fields we will use a special character, `:`, instead of `.` to reference a struct field.

Considering the following stream:
```CREATE STREAM orders (ordertime bigint, orderid bigint, address STRUCT < number bigint, street varchar, city varchar, state varchar, zipcode bigint>) WITH ...```

with address as a struct column we can refer to the fields in a query as follows:

```SELECT address:city, address:state FROM orders;```

Note that if you want to prefix a column with stream name/alias you still need to use `'`. For instance:

``` SELECT orders.address:city, o.address:state FROM orders o;```

### Testing done 
The existing tests were updated to use `:`. Added test for SqlFormatting.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java [hunk 1]
```diff
@@ -148,5 +222,18 @@ public void shouldFormatOuterJoinWithoutJoinWindow() {
...
+
+  @Test
+  public void shouldFormatSelectQueryCorrectly() {
+    final String statementString =
+        "CREATE STREAM S AS SELECT a.address:city FROM address a;";
+    final Statement statement = KSQL_PARSER.buildAst(statementString, metaStore).get(0);
+
+    String s = SqlFormatter.formatSql(statement);
+
+    System.out.println();
```
Comment (apurvam) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java]: What is this testing exactly?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 2]
```diff
@@ -635,12 +632,12 @@ public Node visitSelectSingle(SqlBaseParser.SelectSingleContext context) {
                     dereferenceExpression.getFieldName()
                 )
           )) {
-          alias = Optional.of(
+          alias = Optional.of(replaceDotColon(
```
Comment (apurvam) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Not sure I follow the motivation for this. The argument for `replaceDotColon` is going to be a string without colons, since it is being constructed from two strings without colon, ie. the `base` and the `fieldName`.

File: ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4 [hunk 3]
```diff
@@ -663,11 +664,11 @@ DECIMAL_VALUE
     ;
 
 IDENTIFIER
-    : (LETTER | '_') (LETTER | DIGIT | '_' | '@' | ':')*
+    : (LETTER | '_') (LETTER | DIGIT | '_' | '@' )*
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4]: AS you say, it probably wouldn't work as its an invalid java identifier. Such a break is OK though. We just need to clearly document it.

[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4]: This won't affect the backward compatibility. As you mentioned `:` is not permitted in the java identifier and regardless we cannot use it in identifier name.
I think we should remove `@` for the same reason but this can be done in a separate PR.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java (hunk 1 above)
Comment (hjafarpour) [reply ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java]: Oops, I forgot to complete this one. Updated it.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 2 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: `base` is not a string, it's an expression and can include `:`. I also made more changes here after switching to `->`. Let me know if you have more questions.

File: ksql-engine/src/test/resources/query-validation-tests/more-complex-struct.json [hunk 4]
```diff
@@ -122,31 +122,31 @@
...
           "key": 101,
-          "value": "{\"SL4_DOUBLE_ARRAY\":[74.68834760950283,73.40355939056849,72.15630069159221],\"SL2_STRUCT_MAP\":{\"mapkey\":{\"SL2_3_STRING\":\"SL2_3_95\",\"SL2_3_DOUBLE\":51.92447078198864}}}",
+          "value": "{\"NESTED_STRUCT_FIELD__SL1_STRUCT__SL2_STRUCT__SL3_STRUCT__SL4_DOUBLE_ARRAY\":[74.68834760950283,73.40355939056849,72.15630069159221],\"NESTED_STRUCT_FIELD__SL1_STRUCT__SL2_STRUCT_MAP\":{\"mapkey\":{\"SL2_3_STRING\":\"SL2_3_95\",\"SL2_3_DOUBLE\":51.92447078198864}}}",
           "timestamp": 0
         },
         {
           "topic": "S2",
           "key": 101,
-          "value": "{\"SL4_DOUBLE_ARRAY\":[32.27711275538546,3.7338613605496174,52.41743778835066],\"SL2_STRUCT_MAP\":{\"mapkey\":{\"SL2_3_STRING\":\"SL2_3_1\",\"SL2_3_DOUBLE\":85.05450978718197}}}",
+          "value": "{\"NESTED_STRUCT_FIELD__SL1_STRUCT__SL2_STRUCT__SL3_STRUCT__SL4_DOUBLE_ARRAY\":[32.27711275538546,3.7338613605496174,52.41743778835066],\"NESTED_STRUCT_FIELD__SL1_STRUCT__SL2_STRUCT_MAP\":{\"mapkey\":{\"SL2_3_STRING\":\"SL2_3_1\",\"SL2_3_DOUBLE\":85.05450978718197}}}",
```
Comment (big-andy-coates) [inline ksql-engine/src/test/resources/query-validation-tests/more-complex-struct.json]: Why the change in column names? These new names are pretty horrid. I definitely preferred the previous names.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 5]
```diff
@@ -650,7 +652,7 @@ public Node visitSelectSingle(SqlBaseParser.SelectSingleContext context) {
   }
 
   private String replaceDotColon(String input) {
-    return input.replace(".", "_").replace(":", "__");
+    return input.replace(".", "_").replace("->", "__");
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: The use of `"->"` throughput the code should be using a constant. As should `"."`

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java [hunk 6]
```diff
@@ -381,25 +381,6 @@ public Expression visitFunctionCall(FunctionCall node, Context<C> context) {
       return node;
     }
 
-    @Override
-    protected Expression visitLambdaExpression(LambdaExpression node, Context<C> context) {
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java]: Why have we deleted all this? How does this relate to the PR?

File: ksql-engine/src/test/resources/query-validation-tests/more-complex-struct.json (hunk 4 above)
Comment (hjafarpour) [reply ksql-engine/src/test/resources/query-validation-tests/more-complex-struct.json]: You can easily have conflict in the previous names. In this form you if you select a struct field and don't specify an alias you will have a name created representing the full path of the field. If we used the previous aliasing we would not be able to run valid queries where struct field name is the same as a column name or another struct field name.
Consider the following:

```CREATE STREAM foo (ID BIGINT, ITEM_INFO STRUCT<ID BIGINT>) WITH ...```
```CREATE STREAM bar AS SELECT ID, ITEM_INFO->ID FROM foo;```

Previously the above query would fails since the same alias will be used for two columns, `ID`. But now we will use two aliases: `ID, ITEM_INFO_ID`, and the query won't fail.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java (hunk 6 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java]: We were not supporting Lambda anyways and also since it was using the same characters was confusing the grammar to pick Lambda instead of struct field reference. By removing it we can use the struct field reference correctly and also won't have unused/unsupported feature in the grammar.

**Comment by hjafarpour** [comment]
@big-andy-coates I pushed the previous commit before reading your comments. The next commit includes your latest comments along with a test for the ambiguity case you mentioned.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java (hunk 6 above)
Comment (blueedgenick) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java]: but isn't this kinda rash ? - why are we precluding adding a cool future feature (not saying it should be in the codebase already, just trying to leave space in the plan for it) simply by being opinionated over our choice of separator char ?
FWIW i already gave feedback that not only does a choice of `->` preclude doing something interesting in the future that we may well want to add, it's also confusing to the target audience: folks from a sql and big data background are used to single-character identifiers, whereas the folks from a java-programming background are going to read it as a lamba.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 7]
```diff
@@ -652,6 +654,10 @@ public Node visitSelectSingle(SqlBaseParser.SelectSingleContext context) {
     return new SingleColumn(getLocation(context), selectItemExpression, alias);
   }
 
+  private String replaceDotFieldRef(String input) {
+    return input.replace(DOT, "_").replace(STRUC_FIELD_REF, "__");
```
Comment (rodesai) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Are we worried about this generating ambiguous names? One place this might trip us up is in aggregation.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java [hunk 8]
```diff
@@ -235,7 +234,10 @@ protected String visitSymbolReference(SymbolReference node, Boolean context) {
     @Override
     protected String visitDereferenceExpression(DereferenceExpression node, Boolean unmangleNames) {
       String baseString = process(node.getBase(), unmangleNames);
-      return baseString + "." + formatIdentifier(node.getFieldName());
+      if (node.getBase() instanceof QualifiedNameReference) {
+        return baseString + "." + formatIdentifier(node.getFieldName());
```
Comment (rodesai) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java]: . and -> should be constants somewhere

File: ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java [hunk 9]
```diff
@@ -148,5 +226,17 @@ public void shouldFormatOuterJoinWithoutJoinWindow() {
     final String expected = "left L\nFULL OUTER JOIN right R ON (('left.col0' = 'right.col0'))";
     assertEquals(expected, SqlFormatter.formatSql(join));
   }
+
```
Comment (rodesai) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java]: Should we add a case with a struct dereference here?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java (hunk 6 above)
Comment (rodesai) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java]: I don't think we're precluding it. It just means that we would have to choose a different syntax for it.

I'm also not _that_ worried about confusing java developers. I would think most java developers have some familiarity with c/c++, so the notion of using a '->' to dereference a struct should feel familiar.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java (hunk 9 above)
Comment (hjafarpour) [reply ksql-parser/src/test/java/io/confluent/ksql/parser/SqlFormatterTest.java]: We don't support non-column values in the join criteria yet.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 7 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: In the aggregates we prefix internal columns with `KSQL_AGG_VARIABLE_` so I think it should be easy to differentiate.

File: ksql-engine/src/test/resources/query-validation-tests/more-complex-struct.json (hunk 4 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/resources/query-validation-tests/more-complex-struct.json]: Yer, but on the down side, `NESTED_STRUCT_FIELD__SL1_STRUCT__SL2_STRUCT__SL3_STRUCT__SL4_DOUBLE_ARRAY` is a bloody horrible column name to be outputting ;) . I guess they can use aliasing to be make it more manageable thought...

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java (hunk 6 above)
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/ExpressionTreeRewriter.java]: Already mentioned to Hojjat that we can use '=>' for lambdas.  '->' is used by a couple of other dbs for referencing struct fields.

File: ksql-engine/src/test/resources/query-validation-tests/simple-struct.json [hunk 10]
```diff
@@ -374,6 +374,29 @@
           "timestamp": 0
         }
       ]
+    },
+    {
+      "name": "simple struct select for ambiguity",
```
Comment (big-andy-coates) [inline ksql-engine/src/test/resources/query-validation-tests/simple-struct.json]: perfect!

File: ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java [hunk 11]
```diff
@@ -235,7 +234,10 @@ protected String visitSymbolReference(SymbolReference node, Boolean context) {
...
+      }
+      return baseString + "->" + formatIdentifier(node.getFieldName());
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java]: Should use constant.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java (hunk 8 above)
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java]: should use constant.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 12]
```diff
@@ -652,6 +654,10 @@ public Node visitSelectSingle(SqlBaseParser.SelectSingleContext context) {
     return new SingleColumn(getLocation(context), selectItemExpression, alias);
   }
 
+  private String replaceDotFieldRef(String input) {
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: `static` method, `final` param.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java (hunk 11 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/ExpressionFormatter.java]: Already made them constant as @rodesai suggested too.
